### ISSUE TITLE
Dashboard PTY reattach logs uncaught WebSocketDisconnect during snapshot replay

### ISSUE THREAD

**Issue by kaykayyali** [NousResearch/hermes-agent#63394]
## Bug

The keep-alive PTY route replays buffered output before entering its existing WebSocket disconnect/cleanup block. If the client disconnects during that replay, `send_bytes()` raises `WebSocketDisconnect`, Uvicorn logs an ASGI exception, and the route skips its normal registry detach.

Current `origin/main` (`7b5ba20`):

```python
await session.attach(ws)

try:
    while True:
        msg = await ws.receive()
        ...
except WebSocketDisconnect:
    pass
finally:
    PTY_REGISTRY.detach(attach_token, ws)
```

`PtySession.attach()` performs the vulnerable replay:

```python
snap = self.buffer.snapshot()
if snap:
    await ws.send_bytes(snap)
```

Observed exception path:

```text
pty_ws -> session.attach -> ws.send_bytes(snap)
starlette.websockets.WebSocketDisconnect: 1006
```

## Scope correction

This exception is **nonfatal to Uvicorn** and does not terminate the dashboard service. A service outage initially associated with this traceback was caused by a separate explicit process termination. This issue only tracks the uncaught per-connection exception and missed cleanup.

## Expected behavior

A client disappearing during initial replay should be treated like any other WebSocket disconnect: no ASGI error log, and the registry should detach the socket.

## Minimal fix

Move `await session.attach(ws)` inside the route's existing `try / except WebSocketDisconnect / finally` block. This catches only the expected transport exception and preserves cleanup without adding a broad catch to `PtySession`.

A deterministic regression test can make `session.attach()` raise `WebSocketDisconnect` and assert that `PTY_REGISTRY.detach()` is called.

**Comment by GottZ** [comment]
> *This was generated by AI during triage.*

#### Summary

20 PRs are linked to this issue complex, but only #63403 directly addresses #63394's snapshot-replay disconnect and missed detach; the other 19 concern stale PTY identity, dashboard-user propagation, pet output, readiness, or cross-tab isolation and do not change the vulnerable replay boundary.

#### Related pull requests

- #60745 **[merged]** `unverified` — (+117/-27) — unrelated merged reference implementation: it scopes the server-side PTY registry key by canonical resume target and profile, fixing stale session/profile reattachment rather than catching a disconnect during snapshot replay. It remains relevant because merged commit 921c17af8826724e9dfb468eec9c308499d43447 is the implementation that supersedes several adjacent session-switch PRs.
- #61045 `unverified` — (+23/-5) — unrelated partial session-switch fix: it namespaces the frontend token only by resume and adds separate descendant-selection changes, but it does not move session.attach(ws) into a disconnect/cleanup boundary. Its visible keep_open review identifies missing profile isolation and regression coverage.
- #61107 `unverified` — (+11/-4) — unrelated duplicate session-switch fix: it duplicates #61045's resume-only frontend token namespacing and leaves both profile isolation and #63394's replay exception path untouched.
- #61290 **[closed]** `unverified` — (+125/-2) — unrelated closed duplicate of #60745: it uses raw resume plus profile in a symmetric backend registry key and tests target isolation. It remains relevant as superseded evidence for the stale-PTY root cause, not as a replay-disconnect fix.
- #61320 **[closed]** `unverified` — (+5/-2) — unrelated closed duplicate of #60745: it adds resume and profile to the registry key, but uses post-fallback resume and lacks the required regressions. It remains relevant because the keep_open review documents why that variant could break default-chat reattachment.
- #61968 `unverified` — (+61/-1) — unrelated pet-output fix: its final diff gates _pet_start_anim() before creating the prompt_toolkit invalidation thread and adds a non-TTY regression test. It does not touch PTY snapshot replay or WebSocket cleanup.
- #62058 `unverified` — (+25/-0) — unrelated partial session-switch fix: it rotates the frontend token when resume changes, but omits profile changes and does not alter the server replay boundary. Its keep_open review confirms the stale-token premise while requiring broader target isolation and tests.
- #62183 **[closed]** `unverified` — (+100/-27) — unrelated closed duplicate of #60745: it repeats raw-resume/profile backend keying with same-target, different-target, and fallback tests. It remains relevant only as superseded session-switch coverage.
- #62563 **[closed]** `unverified` — (+82/-21) — unrelated closed identity-propagation attempt: it carries ticket metadata toward AIAgent, but accepts user_id through a caller-controlled /api/ws query and lets an RPC value win via setdefault; it also leaves a console_ws tuple mismatch. It remains relevant as an unsafe predecessor to #62629, not as a #63394 fix.
- #62622 `unverified` — (+21/-1) — unrelated defective readiness attempt: it queues normal Ink terminal diffs behind a new ptyReady flag but provides no caller for setPtyReady(), so output can remain queued indefinitely. The contributor keep_open review also identifies the actual referenced issue path as PTY registry identity, not this readiness gate.
- #62629 **[closed]** `unverified` — (+397/-24) — unrelated closed identity-propagation attempt: its updated diff reaches the existing gateway and AIAgent, but user_id is still read from a caller-controlled query and setdefault permits an RPC override. It remains relevant as the most complete #62549 attempt, while its visible keep_open review concerned that separate cross-process identity boundary.
- #62877 **[closed]** `unverified` — (+23/-13) — unrelated closed partial duplicate of #62563: it exports ticket-derived identity only to the PTY child, which does not reach AIAgent in the normal already-running gateway path, and it leaves console_ws on the old tuple contract. It remains relevant as evidence that child-environment propagation alone is insufficient.
- #62903 `unverified` — (+6/-1) — unrelated wrong-layer pet fix: it suppresses every non-TTY Ink terminal diff while leaving the reported classic-CLI pet animation thread running. The contributor keep_open review therefore finds no repair for either #61964's actual path or #63394.
- #63217 **[closed]** `unverified` — (+34/-15) — unrelated closed partial identity duplicate: it correctly extracts ticket identity and updates tuple consumers, but only exports HERMES_SESSION_USER_ID to the PTY child rather than crossing into the existing gateway's AIAgent construction. It remains relevant as another superseded env-only attempt.
- #63403 `best fix` — (+43/-2) — best existing fix: it moves session.attach(ws) inside the route's existing WebSocketDisconnect/finally boundary, so a replay disconnect is handled and PTY_REGISTRY.detach() still runs; its deterministic test makes attach raise and asserts detach. This is consistent with the visible keep_open review, which confirms the exact current-main exception path and rates the focused route-boundary change highly salvageable.
- #63803 **[closed]** `unverified` — (+290/-26) — unrelated closed partial cross-tab fix: it moves attachment ownership toward per-tab storage, but its 75 ms BroadcastChannel ownership timeout can retain a cloned token after a delayed response. It remains relevant because its blocking contributor reviews and self-close document why the cross-tab portion of #62726 is not safely resolved.
- #64163 `unverified` — (+227/-29) — unrelated frontend session/profile fix: it scopes token storage by profile and resume and adds focused tests, but it neither catches replay disconnects nor guarantees detach. Its keep_open review supports that separate stale-PTY mechanism; merged #60745 now enforces the broader invariant at the backend.
- #65313 `unverified` — (+6/-2) — unrelated incomplete backend session-switch fix: it keys on post-fallback resume only, omits profile, and can create a second default-chat key after the active-session file is populated. Its keep_open review explicitly requires raw resume, profile, and regression coverage.
- #65473 **[closed]** `unverified` — (+34/-0) — unrelated closed redundant session-switch fix: it rotates the frontend token on resume changes and includes a separate resize adjustment. It remains relevant because the automated close verdict identifies the titled behavior as already implemented by 921c17af8826724e9dfb468eec9c308499d43447 at hermes_cli/web_server.py:16283-16290, with tests/test_pty_keepalive_ws.py:78-87 covering distinct resume targets.
- #67217 **[closed]** `unverified` — (+16/-4) — unrelated closed frontend duplicate superseded by #60745: it scopes localStorage by raw profile and resume, but can split parent and canonical-child identities. It remains relevant because the author closed it after confirming the merged backend implementation already provides canonical target isolation.

#### Duplicates

Session-switch duplicates: #60745, #61290, #61320, and #62183 implement the same backend resume/profile registry-key approach; #61045 and #61107 are resume-scoped frontend twins, with #64163 and #67217 broader profile+resume variants; #62058 and #65473 both rotate the frontend token on resume changes. Identity-propagation duplicates: #62563, #62629, #62877, and #63217 overlap on ticket metadata and PTY/gateway plumbing, with #62877 and #63217 stopping at the child environment. #63403 has no duplicate in the supplied PR list; its same-race alternative #63193 is referenced but not included.

#### Suggested consolidation

Keep #63403 open with a salvage path as the best existing fix: rebase its two-file route-boundary change onto current main, preserve the narrow WebSocketDisconnect catch and unconditional PTY_REGISTRY.detach() assertion, then rerun the focused keep-alive WebSocket tests. This follows the visible keep_open review rather than deviating from it: the review confirms both the vulnerable PtySession.attach() replay and the adequacy of the handler-level regression. Do not consolidate the unrelated PTY-identity, auth-propagation, pet, readiness, or cross-tab PRs into #63403; the closed duplicates can remain closed, while the still-open unrelated candidates should be triaged within their own issue complexes.

#### Complex graph

```mermaid
flowchart TD
    classDef open fill:#dbeafe,stroke:#1d4ed8,color:#1e3a8a
    classDef merged fill:#dcfce7,stroke:#15803d,color:#14532d
    classDef closed fill:#e5e7eb,stroke:#6b7280,color:#1f2937
    classDef unverified fill:#f3f4f6,stroke:#9ca3af,color:#374151
    classDef best stroke-width:3px,stroke:#b45309
    classDef target stroke-width:3px,stroke:#4338ca
    I61284(["issue #61284 (closed)"])
    I61964(["issue #61964 (open)"])
    I62549(["issue #62549 (open)"])
    I63394(["issue #63394 (open)"])
    I64154(["issue #64154 (open)"])
    P61968["PR #61968 (open)"]
    P62058["PR #62058 (open)"]
    P62622["PR #62622 (open)"]
    P63403["PR #63403 (open)"]
    subgraph Dup60745 ["PRs duplicating each other"]
        P60745["PR #60745 (merged)"]
        P61290["PR #61290 (closed)"]
        P61320["PR #61320 (closed)"]
        P62183["PR #62183 (closed)"]
    end
    subgraph Dup61045 ["PRs duplicating each other"]
        P61045["PR #61045 (open)"]
        P61107["PR #61107 (open)"]
    end
    subgraph Dup62563 ["PRs duplicating each other"]
        P62563["PR #62563 (closed)"]
        P62629["PR #62629 (closed)"]
    end
    More["+8 more PRs not shown"]
    MoreI(["+1 more issues not shown"])
    P60745 -.->|unverified| I63394
    P61045 -.->|unverified| I63394
    P61107 -.->|unverified| I63394
    P61290 -.->|unverified| I63394
    P61320 -.->|unverified| I63394
    P61968 -.->|unverified| I63394
    P62058 -.->|unverified| I63394
    P62183 -.->|unverified| I63394
    P62563 -.->|unverified| I63394
    P62622 -.->|unverified| I63394
    P62629 -.->|unverified| I63394
    P63403 -->|best fix| I63394
    P60745 -->|best fix| I61284
    P61045 -.->|partial| I61284
    P61107 -.->|partial| I61284
    P61290 -->|fixes| I61284
    P61320 -.->|partial| I61284
    P62058 -.->|partial| I61284
    P62183 -->|fixes| I61284
    P61968 -->|best fix| I61964
    P62563 -.->|partial| I62549
    P62629 -->|best fix| I62549
    P60745 -->|best fix| I64154
    class I61284 closed
    class I61964 open
    class I62549 open
    class I63394 open
    class I64154 open
    class I63394 target
    class P60745 merged
    class P61045 open
    class P61107 open
    class P61290 closed
    class P61320 closed
    class P61968 open
    class P62058 open
    class P62183 closed
    class P62563 closed
    class P62622 open
    class P62629 closed
    class P63403 open
    class P60745 best
    class P60745 best
    class P61968 best
    class P62629 best
    class P63403 best
    class More closed
    class MoreI closed
    click I61284 "https://github.com/NousResearch/hermes-agent/issues/61284"
    click I61964 "https://github.com/NousResearch/hermes-agent/issues/61964"
    click I62549 "https://github.com/NousResearch/hermes-agent/issues/62549"
    click I63394 "https://github.com/NousResearch/hermes-agent/issues/63394"
    click I64154 "https://github.com/NousResearch/hermes-agent/issues/64154"
    click P60745 "https://github.com/NousResearch/hermes-agent/pull/60745"
    click P61045 "https://github.com/NousResearch/hermes-agent/pull/61045"
    click P61107 "https://github.com/NousResearch/hermes-agent/pull/61107"
    click P61290 "https://github.com/NousResearch/hermes-agent/pull/61290"
    click P61320 "https://github.com/NousResearch/hermes-agent/pull/61320"
    click P61968 "https://github.com/NousResearch/hermes-agent/pull/61968"
    click P62058 "https://github.com/NousResearch/hermes-agent/pull/62058"
    click P62183 "https://github.com/NousResearch/hermes-agent/pull/62183"
    click P62563 "https://github.com/NousResearch/hermes-agent/pull/62563"
    click P62622 "https://github.com/NousResearch/hermes-agent/pull/62622"
    click P62629 "https://github.com/NousResearch/hermes-agent/pull/62629"
    click P63403 "https://github.com/NousResearch/hermes-agent/pull/63403"
```
*Graph: solid arrow = fixes / best fix, dashed arrow = partial or unverified (see edge label); boxed group = PRs duplicating each other; amber border = best fix; indigo border = target; gray node = closed (state tag in the node label).*

_Cross-PR triage: Reviewed 20 pull requests and 6 issues in this complex. Each diff was read against this issue; Assessment working set: 127 kB of PR diffs, 62 kB of issue/PR text, 86 kB of discussion (76 comments), 27 verify verdicts. verdicts reflect diff content, not PR titles. Part of an automated triage batch._
